### PR TITLE
sampling: add trace group sampling reservoirs

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Introduce a type for maintaining a collection of trace groups.

For each trace group we measure the observed ingest rate for
the trace group (number of root traces per sampling interval),
and maintain a weighted random sampling reservoir for recording
trace IDs. Trace IDs are weighted by root transaction duration,
which means slower traces will have a greater chance of being
captured.

Metrics/monitoring for this will come later.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
- [x] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

`make test`

## Related issues

None.

### CODE REVIEW

File: x-pack/apm-server/sampling/groups.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		if group.reservoir.Size() == minReservoirSize {
+			if group.total == 0 && len(g.groups) == g.maxGroups {
+				delete(g.groups, key)
+			}
+		}
+	}
+	return traceIDs
+}
+
+// finalizeSampledTraces locks the group, appends its current trace IDs to
```
Comment (axw) [inline x-pack/apm-server/sampling/groups.go:153]: it does not lock the group - I updated the code and forgot to update this, will fix

File: x-pack/apm-server/sampling/groups.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+var errTooManyTraceGroups = errors.New("too many trace groups")
+
+// traceGroups maintains a collection of trace groups.
+type traceGroups struct {
+	// defaultSamplingFraction holds the default sampling fraction to
+	// set for new trace groups. See traceGroup.samplingFraction.
+	defaultSamplingFraction float64
+
+	// ingestRateCoefficient is λ, the coefficient for calculating
```
Comment (axw) [inline x-pack/apm-server/sampling/groups.go:27]: perhaps "ingestRateDecayFactor" would be a more helpful name

File: x-pack/apm-server/sampling/groups.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+// finalizeSampledTraces locks the groups, appends their current trace IDs to
+// traceIDs, and returns the extended slice. On return the groups' sampling
+// reservoirs will be reset.
+func (g *traceGroups) finalizeSampledTraces(traceIDs []string) []string {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	for key, group := range g.groups {
+		traceIDs = group.finalizeSampledTraces(traceIDs, g.ingestRateCoefficient)
+		if group.reservoir.Size() == minReservoirSize {
+			if group.total == 0 && len(g.groups) == g.maxGroups {
```
Comment (axw) [inline x-pack/apm-server/sampling/groups.go:145]: - group.total is always zero after finalizeSampledTraces
- It was meant to be ||

File: x-pack/apm-server/sampling/groups.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		g.ingestRate *= 1 - ingestRateDecayFactor
+		g.ingestRate += ingestRateDecayFactor * float64(g.total)
+	}
+	desiredTotal := int(math.Round(g.samplingFraction * float64(g.total)))
+	g.total = 0
+
+	for n := g.reservoir.Len(); n > desiredTotal; n-- {
+		// The reservoir is larger than the desired fraction of the
+		// observed total number of traces in this interval. Pop the
+		// lowest weighed traces to limit to the desired total.
```
Comment (axw) [inline x-pack/apm-server/sampling/groups.go:174]:
```suggestion
		// lowest weighted traces to limit to the desired total.
```

File: x-pack/apm-server/sampling/groups.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+// minimum reservoir size (low ingest or sampling rate) may be removed. These
+// groups may also be removed if they have seen no activity in this interval.
+func (g *traceGroups) finalizeSampledTraces(traceIDs []string) []string {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	maxGroupsReached := len(g.groups) == g.maxGroups
+	for key, group := range g.groups {
+		total := group.total
+		traceIDs = group.finalizeSampledTraces(traceIDs, g.ingestRateDecayFactor)
+		if n := group.reservoir.Size(); n == minReservoirSize {
```
Comment (simitt) [inline x-pack/apm-server/sampling/groups.go:149]: nit: `if group.reservoir.Size() == minReservoirSize` (you are not using `n` anywhere else)

**Review by simitt — APPROVED** [review]
Sorry for the late review; and thanks for all the comments - very helpful!

**Pull request merged**